assets/js/modules/analytics/datastore/report.js

```diff
@@ -0,0 +1,144 @@
+import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
+import { isValidDateRange, isValidOrders } from '../../../util/report-validation';
+import { isValidDimensions, isValidMetrics } from '../util/report-validation';
+
+const fetchGetReportStore = createFetchStore( {
+	baseName: 'getReport',
+	controlCallback: ( { options } ) => {
+		return API.get( 'modules', 'analytics', 'report', options );
+	},
+	reducerCallback: ( state, report, { options } ) => {
+		return {
+			...state,
+			reports: {
+				...state.reports,
+				[ stringifyObject( options ) ]: report,
+			},
+		};
+	},
+	argsToParams: ( options ) => {
+		const {
+			metrics,
+			dimensions,
+			orderby,
+		} = options;
+
+		invariant( isPlainObject( options ), 'Options for Analytics report must be an object.' );
+		invariant( isValidDateRange( options ), 'Either date range or start/end dates must be provided for Analytics report.' );
+		invariant(
+			isValidMetrics( metrics ),
+			'Metrics for an Analytics report must be either a string, an array of strings, an object, an array of objects or a mix of strings and objects. If an object is used, it must have "expression" and "alias" properties.',
+		);
+
+		if ( dimensions ) {
+			invariant(
+				isValidDimensions( dimensions ),
+				'Dimensions for an Analytics report must be either a string, an array of strings, an object, an array of objects or a mix of strings and objects. If an object is used, it must have "name" property.',
+			);
+		}
+
+		if ( orderby ) {
+			invariant(
+				isValidOrders( orderby ),
+				'Orders for an Analytics report must be either an object or an array of objects where each object should have "fieldName" and "sortOrder" properties.',
+			);
+		}
+
+		return { options };
+	},
+} );
+
+const BASE_INITIAL_STATE = {
+	reports: {},
+};
+
+const baseResolvers = {
+	*getReport( options = {} ) {
+		const registry = yield Data.commonActions.getRegistry();
+		const existingReport = registry.select( STORE_NAME ).getReport( options );
+
+		// If there are already alerts loaded in state, consider it fulfilled
+		// and don't make an API request.
+		if ( existingReport ) {
+			return;
+		}
+
+		yield fetchGetReportStore.actions.fetchGetReport( options );
+	},
+};
+
+const baseSelectors = {
+	/**
+	 * Gets an Analytics report for the given options.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object}         state                       Data store's state.
+	 * @param {Object}         options                     Options for generating the report.
+	 * @param {string}         options.startDate           Required, unless dateRange is provided. Start date to query report data for as YYYY-mm-dd.
+	 * @param {string}         options.endDate             Required, unless dateRange is provided. End date to query report data for as YYYY-mm-dd.
+	 * @param {string}         options.dateRange           Required, alternative to startDate and endDate. A date range string such as 'last-28-days'.
```

Comment:
Maybe this should be optional? Since it's only required as an alternative to `startDate` and `endDate`. Not sure, but I guess technically one of them is optional, and I think since this only here for BC, we should mark it as optional.

Alternatively, maybe we just shouldn't document the three legacy arguments at all? We need them for our own BC, but we don't really want other people to use them, so 🤷‍♂️ ? Previously mentioned in https://github.com/google/site-kit-wp/issues/1775#issuecomment-658236590

Reply:
I would prefer to leave it here and we can't make it optional because it is required if other dates are not provided. If you still think we need to remove it or make it optional, let me know and i'll update it.